File: app/src/main/java/com/ewaldmire/osmride/ui/ride/ElevationProfileChart.kt
package com.ewaldmire.osmride.ui.ride

import androidx.compose.foundation.Canvas
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Path
import com.ewaldmire.osmride.route.Route

private data class ElevationPoint(val distanceMeters: Double, val elevationMeters: Double)

/**
 * Elevation-vs-distance strip along the route, in the same visual language as
 * WorkoutProfileChart: a filled area under the elevation line, with a vertical marker for how
 * far along [progressMeters] the rider currently is. Doubles as the ride's overall progress
 * indicator (a separate LinearProgressIndicator next to it would just show the same position
 * twice) - routes with no elevation data (e.g. a GPX with no <ele> tags) fall back to a flat
 * background so the marker is still visible instead of the whole composable disappearing.
 */
@Composable
fun ElevationProfileChart(route: Route, progressMeters: Double?, modifier: Modifier = Modifier) {
    val points = remember(route.id) {
        route.points.mapNotNull { p -> p.elevationMeters?.let { ElevationPoint(p.cumulativeDistanceMeters, it) } }
    }

    val minElevation = remember(points) { points.minOfOrNull { it.elevationMeters } ?: 0.0 }
    val maxElevation = remember(points) { (points.maxOfOrNull { it.elevationMeters } ?: 1.0).coerceAtLeast(minElevation + 1.0) }
    val totalDistance = route.totalDistanceMeters.coerceAtLeast(1.0)
    val backgroundColor = MaterialTheme.colorScheme.surfaceVariant
    val fillColor = MaterialTheme.colorScheme.primary
    val markerColor = MaterialTheme.colorScheme.onSurface

    Canvas(modifier = modifier) {
        drawRect(color = backgroundColor)

        if (points.size >= 2) {
            val path = Path().apply {
                moveTo(0f, size.height)
                points.forEach { point ->
                    val x = (point.distanceMeters / totalDistance).toFloat() * size.width
                    val fraction = ((point.elevationMeters - minElevation) / (maxElevation - minElevation)).toFloat()
                    lineTo(x, size.height - fraction * size.height)
                }
                lineTo(size.width, size.height)
                close()
            }
            drawPath(path, color = fillColor.copy(alpha = 0.5f))
        }

        if (progressMeters != null) {
            val x = (progressMeters.coerceIn(0.0, totalDistance) / totalDistance).toFloat() * size.width
            drawLine(color = markerColor, start = Offset(x, 0f), end = Offset(x, size.height), strokeWidth = 5f)
        }
    }
}

File: linux-client/osm_ride_linux/ui/elevation_profile_chart.py
"""Elevation-vs-distance strip along the route, in the same visual language as
WorkoutProfileChart: a filled area under the elevation line, with a vertical marker for how far
along the route the rider currently is. Doubles as the ride's overall progress indicator (a
separate progress bar would be redundant with the marker line this already draws) - routes with
no elevation data (e.g. a GPX with no <ele> tags) fall back to a flat baseline so the marker is
still visible instead of the whole widget disappearing.

Mirrors app/src/main/java/com/ewaldmire/osmride/ui/ride/ElevationProfileChart.kt's Canvas
drawing, using Cairo (GTK's drawing API) instead of Compose's Canvas.
"""

from __future__ import annotations

import gi

gi.require_version("Gtk", "4.0")
from gi.repository import Gtk  # noqa: E402

from ..route.models import Route  # noqa: E402


class ElevationProfileChart(Gtk.DrawingArea):
    def __init__(self) -> None:
        super().__init__()
        self._points: list[tuple[float, float]] = []  # (cumulative_distance_meters, elevation_meters)
        self._total_distance_meters = 1.0
        self._progress_meters: float | None = None
        self.set_size_request(-1, 22)
        self.set_draw_func(self._on_draw)

    def set_route(self, route: Route | None) -> None:
        if route is None:
            self._points = []
        else:
            self._points = [
                (p.cumulative_distance_meters, p.elevation_meters) for p in route.points if p.elevation_meters is not None
            ]
            self._total_distance_meters = max(route.total_distance_meters, 1.0)
        self.queue_draw()

    def set_progress_meters(self, meters: float | None) -> None:
        self._progress_meters = meters
        self.queue_draw()

    def _on_draw(self, _area: Gtk.DrawingArea, cr, width: int, height: int) -> None:  # noqa: ANN001 - cairo.Context
        cr.set_source_rgb(0.85, 0.85, 0.85)
        cr.rectangle(0, 0, width, height)
        cr.fill()

        if len(self._points) >= 2:
            elevations = [e for _, e in self._points]
            min_elevation = min(elevations)
            max_elevation = max(max(elevations), min_elevation + 1.0)

            cr.set_source_rgba(0.13, 0.44, 0.71, 0.5)
            cr.move_to(0, height)
            for distance, elevation in self._points:
                x = (distance / self._total_distance_meters) * width
                fraction = (elevation - min_elevation) / (max_elevation - min_elevation)
                cr.line_to(x, height - fraction * height)
            cr.line_to(width, height)
            cr.close_path()
            cr.fill()

        if self._progress_meters is not None:
            x = min(max(self._progress_meters, 0.0), self._total_distance_meters) / self._total_distance_meters * width
            cr.set_source_rgb(0.1, 0.1, 0.1)
            cr.set_line_width(2)
            cr.move_to(x, 0)
            cr.line_to(x, height)
            cr.stroke()
